working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,913 @@
+
+*   The upwards inference key type of an `if` map element without an `else` is
+    the key type of the "then" element, likewise for the value type.
+
+*   The upwards inference key type of an `if-else` map element is the least
+    upper bound of the key types of the "then" and "else" elements, likewise for
+    the value type.
+
+*   The upwards inference key type of a `for` map element is the key type of the
+    body element, likewise for the value type.
+
+### Type promotion
+
+As with the `if` statement, the condition expression of an `if` element induces
+type promotion in the "then" element of the `if` when the condition expression
+shows that a variable has some type and promotion isn't otherwise aborted.
+
+### Const collections
+
+An `if` element can be used in a const collection, provided its condition
+expression and body elements are all const.
+
+`for` elements cannot be used in const lists and maps.
+
+## Dynamic Semantics
+
+The new dynamic semantics are a superset of the original behavior. To avoid
+redundancy and handle nested uses, the semantics are expressed in terms of a
+separate procedure below:
+
+### Lists
+
+1.  Create a fresh instance of `list` of a class that implements `List<E>`.
+
+    An implementation is, of course, free to optimize pre-allocate a list of the
+    correct capacity when its size is statically known. Note that when spread
+    arguments come into play, it's no longer always possible to statically tell
+    the final size of the resulting flattened list.
+
+1.  For each `element` in the list literal:
+
+    1.  Evaluate `element` using the procedure below.
+
+1.  The result of the literal expression is `list`.
+
+### Maps
+
+A map literal of the form `<K, V>{entry_1 ... entry_n}` is evaluated as follows:
+
+1.  Allocate a fresh instance `map` of a class that implements `LinkedHashMap<K,
+    V>`.
+
+1.  For each `element` in the map literal:
+
+    1.  Evaluate `element` using the procedure below.
+
+1.  The result of the map literal expression is `map`.
+
+### To evaluate a collection `element`:
+
+This procedure handles elements in both list and map literals because the only
+difference is how an atomic list element or map entry is handled. The control
+flow parts are the same so are unified here.
+
+1.  If `element` is an `if` element:
+
+    1.  Evaluate the condition expression to a value `condition`.
+
+    1.  Subject `condition` to boolean conversion to a value `result`.
+
+    1.  If `result` is `true`:
+
+        1.  Evaluate the "then" element using this procedure.
+
+    1.  Else, if there is an "else" element of the `if`:
+
+        1.  Evaluate the "else" element using this procedure.
+
+1.  Else, if `element` is a `for-in` element:
+
+    1.  Evaluate the iterator expression to a value `sequence`.
+
+    1.  Evaluate `sequence.iterator` to a value `iterator`.
+
+    1.  Loop:
+
+        1.  If `iterator.moveNext()` returns `false`, exit the loop.
```

Comment:
Here you also need boolean conversion/checking. Even if `sequence` implements `Iterable` (and I'm not sure you've said that it should), it can return `null`.

In practice, we will likely abstract the current semantics of a `for` statement over the "body", so we can use the same specification for `for` statements and `for` literal elements.
That should handle all the book-keeping of for loops. We do want them to act exactly the same, so it's safer to avoid spec duplication :)

Reply:
Done.

Unifying the spec for `for` statements and `for` elements sounds great.